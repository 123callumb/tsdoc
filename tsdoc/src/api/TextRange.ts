/**
 * Text coordinates represented as a line number and column number.
 *
 * @remarks
 * The first character in a file is considered to be in column 1 of line 1.
 * The location with column 0 and line 0 is used to represent an empty, unspecified,
 * or unknown location.
 */
export interface ITextLocation {
  line: number;
  column: number;
}

/**
 * Efficiently references a range of text from a string buffer.
 */
export class TextRange {
  /**
   * Used to represent an empty or unknown range.
   */
  public static readonly empty: TextRange = new TextRange('', 0, 0);

  /**
   * The starting index into the associated text buffer.
   */
  public readonly pos: number;

  /**
   * The (non-inclusive) ending index for the associated text buffer.
   * For example if the string is called `text`, then the range would correspond
   * to `text.substring(pos, end)`.
   */
  public readonly end: number;

  /**
   * The string buffer that the `pos` and `end` indexes refer to.
   */
  public readonly buffer: string;

  /**
   * Constructs a TextRange that corresponds to an entire string object.
   */
  public static fromString(buffer: string): TextRange {
    return new TextRange(buffer, 0, buffer.length);
  }

  /**
   * Constructs a TextRange that corresponds to an entire string object.
   */
  public static fromStringRange(buffer: string, pos: number, end: number): TextRange {
    return new TextRange(buffer, pos, end);
  }

  /**
   * Constructs a TextRange that corresponds to a different range of an existing buffer.
   */
  public getNewRange(pos: number, end: number): TextRange {
    return new TextRange(this.buffer, pos, end);
  }

  /**
   * Returns the range from the associated string buffer.
   */
  public toString(): string {
    return this.buffer.substring(this.pos, this.end);
  }

  /**
   * Calculates the line and column number for the specified offset into the buffer.
   *
   * @remarks
   * This is a potentially expensive operation.
   *
   * @param offset - an integer offset
   * @param buffer - the buffer
   */
  public getLocation(offset: number): ITextLocation {
    // TODO: Consider caching or optimizing this somehow
    let line: number = 1;
    let column: number = 1;

    let index: number = 0;

    if (offset >= 0 && offset < this.buffer.length) {
      while (index < offset) {
        const c: string = this.buffer[index];
        ++index;

        if (c === '\r') {
          continue;
        }

        if (this.buffer[index] === '\n') {
          ++line;
          column = 1;
          continue;
        }

        ++column;
      }

      return { line: line, column: column };
    }

    // No match
    return { line: 0, column: 0 };
  }

  private constructor(buffer: string, pos: number, end: number) {
    this.buffer = buffer;
    this.pos = pos;
    this.end = end;
    this._validateBounds();
  }

  private _validateBounds(): TextRange {
    if (this.pos < 0) {
      throw new Error('TextRange.pos cannot be negative');
    }
    if (this.end < 0) {
      throw new Error('TextRange.end cannot be negative');
    }
    if (this.end < this.pos) {
      throw new Error('TextRange.end cannot be smaller than TextRange.pos');
    }
    if (this.pos > this.buffer.length) {
      throw new Error('TextRange.pos cannot exceed the associated text buffer length');
    }
    if (this.end > this.buffer.length) {
      throw new Error('TextRange.end cannot exceed the associated text buffer length');
    }

    return this;
  }
}
