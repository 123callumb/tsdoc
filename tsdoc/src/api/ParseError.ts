import { TextRange } from './TextRange';
import { TextBuffer, ITextLocation } from '../internal/TextBuffer';

/**
 * An Error subclass used to report errors that occur while parsing an input.
 */
export class ParseError extends Error {
  public readonly range: TextRange;
  public readonly buffer: string;
  public readonly unformattedMessage: string;
  public readonly innerError: Error | undefined;

  /**
   * Generates a line/column prefix.  Example with line=2 and column=5
   * and message="An error occurred":
   * ```
   * "(2,5): An error occurred"
   * ```
   */
  private static _formatMessage(message: string, range: TextRange, buffer: string): string {
    if (!message) {
      message = 'An unknown error occurred';
    }

    if (range.pos !== 0 || range.end !== 0) {
      const location: ITextLocation = TextBuffer.getLocation(range.pos, buffer);
      if (location.line) {
        return `(${location.line},${location.column}): ` + message;
      }
    }
    return message;
  }

  public constructor(message: string, range: TextRange, buffer: string, innerError?: Error) {
    super(ParseError._formatMessage(message, range, buffer));

    // Boilerplate for extending a system class
    //
    // tslint:disable-next-line:max-line-length
    // https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
    //
    // IMPORTANT: The prototype must also be set on any classes which extend this one
    (this as any).__proto__ = ParseError.prototype; // tslint:disable-line:no-any

    this.unformattedMessage = message;

    this.range = range;
    this.buffer = buffer;
    this.innerError = innerError;
  }
}
