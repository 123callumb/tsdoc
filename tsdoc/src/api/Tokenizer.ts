import { TextRange } from './TextRange';
import { Character } from '../internal/Character';

/**
 * Distinguishes different types of Token objects.
 */
export enum TokenKind {
  /**
   * A token representing the end of the input.
   */
  EndOfInput,
  /**
   * A token representing a sequence of plain text with no special meaning.
   */
  PlainText,
  /**
   * A token representing a virtual newline.  The Token.range will be an empty range,
   * because the actual newline character may be noncontiguous or nonexistent.
   */
  Newline,

  /**
   * A single character that can be used as general punctuation.
   * @remarks
   * The Token.range will always be a string of length 1.
   * When a backslash is interpreted as a literal backslash, it will be returned as
   * PunctuationCharacter instead of BackslashEscapedCharacter.
   */
  PunctuationCharacter,

  /**
   * A backslash followed by another punctuation character.  This disables any special meaning
   * the other character may have (if any).  The backslash itself should not be rendered.
   * @remarks
   * The Token.range will always be a string of length 2.
   */
  BackslashEscapedCharacter
}

/**
 * Represents a contiguous range of characters extracted from one of the doc comment lines
 * being processed by the Tokenizer.  There is a token representing a newline, but otherwise
 * a single token cannot span multiple lines.
 */
export class Token {
  public readonly kind: TokenKind;
  public readonly range: TextRange;
  public readonly line: TextRange;

  public constructor(kind: TokenKind, range: TextRange, line: TextRange) {
    this.kind = kind;
    this.range = range;
    this.line = line;
    if (range.pos < line.pos || range.end > line.end) {
      // Sanity check
      throw new Error('Wrong line');
    }
  }

  public toString(): string {
    return this.range.toString();
  }
}

/**
 * The ICharacter structure uses an empty string to represent the end of the input.
 */
const EOI_CHARACTER: string = '';

/**
 * This helps the Tokenizer keep track of indexes needed to accurately calculated Token.range,
 * given that the buffer index can skip characters when advancing to the next line inside
 * a documentation comment.
 */
interface ICharacter {
  /**
   * Either EOI_CHARACTER (an empty string) to indicate that the end of the input has been reached,
   * or else a string of length one representing a single character extracted from the input buffer.
   */
  value: string;

  /**
   * The buffer index of the character.  This is an index into Tokenizer._buffer.
   */
  index: number;

  /**
   * The input line that this character came from.
   */
  line: TextRange;
}

export class Tokenizer {
  /**
   * The array of doc comment lines being tokenized.
   * These ranges do not include doc comment delimiters such as "/*" or "*".
   */
  public readonly lines: TextRange[];

  // Index into the lines array, indicating the current line being processed.
  private _linesIndex: number;

  // Index into Tokenizer._cachedBuffer, which is the same as this.lines[i].buffer.
  private _bufferIndex: number;

  // If we've reached the end of the input, this is the final Token object.
  private _endToken: Token | undefined;

  // Storage for Token.peekCharacter()
  private _peekedCharacter: ICharacter | undefined;

  // A temporary state used when generating a TokenKind.Newline token
  private _injectedNewline: boolean;

  // To improve performance, this is a cached storage of this.lines[this._linesIndex].
  private _cachedCurrentLine: TextRange | undefined;

  // To improve performance, this is a cached storage of this.lines[i].buffer for all values of i;
  // the lines are assumed to all come from a common buffer.
  private readonly _cachedBuffer: string;

  public constructor(lines: TextRange[]) {
    this.lines = lines;
    this._linesIndex = 0;

    if (this.lines.length === 0) {
      this._cachedCurrentLine = undefined;
      this._cachedBuffer = '';

      this._bufferIndex = 0;
      this._endToken = new Token(TokenKind.EndOfInput, TextRange.empty, TextRange.empty);
    } else {
      this._cachedCurrentLine = this.lines[0];
      this._cachedBuffer = this.lines[0].buffer;

      this._bufferIndex = this._cachedCurrentLine.pos;
      this._endToken = undefined;
    }

    this._peekedCharacter = undefined;
    this._injectedNewline = false;
  }

  /**
   * Extracts and returns the next token from the input stream.
   * @remarks
   * When the end of the input is reached, getToken() will (repeatedly) return
   * the TokenKind.EndOfInput token.
   */
  public getToken(): Token {
    const startCharacter: ICharacter = this._getCharacter();

    switch (startCharacter.value) {
      case EOI_CHARACTER:
        return this._endToken!;

      case '\n':
        return new Token(TokenKind.Newline,
          TextRange.fromStringRange(this._cachedBuffer, startCharacter.index, startCharacter.index),
          startCharacter.line);

      case '\\':
        const next: ICharacter = this._peekCharacter();
        // CommonMark Spec: "Any ASCII punctuation character may be backslash-escaped.'
        // Backslashes before other characters are treated as literal backslashes."
        if (Character.isPunctuation(next.value)) {
          this._getCharacter(); // extract next

          return new Token(TokenKind.BackslashEscapedCharacter,
            TextRange.fromStringRange(this._cachedBuffer, startCharacter.index, next.index + 1),
            startCharacter.line);
        }
        // Treat it as a literal backslash
        return new Token(TokenKind.PunctuationCharacter,
          TextRange.fromStringRange(this._cachedBuffer, startCharacter.index, startCharacter.index + 1),
          startCharacter.line);

      default:
        const lastCharacter: ICharacter = this._skipUntil(['\n']) || startCharacter;
        return new Token(TokenKind.PlainText,
          TextRange.fromStringRange(this._cachedBuffer, startCharacter.index, lastCharacter.index + 1),
          startCharacter.line);
    }
  }

  /**
   * Advances the stream pointer until one of the specified ending characters is reached,
   * or until the end of the input is reached.
   * @returns the last skipped character, or undefined if no characters were skipped
   */
  private _skipUntil(endingCharacters: string[]): ICharacter | undefined {
    let lastSkipped: ICharacter | undefined = undefined;
    while (true) {
      const character: ICharacter = this._peekCharacter();

      if (character.value === EOI_CHARACTER) {
        return lastSkipped;
      }
      if (endingCharacters.indexOf(character.value) >= 0) {
        return lastSkipped;
      }
      lastSkipped = this._getCharacter();
    }
  }

  /**
   * Extracts and returns the next character from the input lines.
   * After each input line is processed, a virtual newline character is returned.
   * When the end of the input is reached, an empty string is returned.
   * The length of the returned string will always be 0 or 1.
   */
  private _getCharacter(): ICharacter {
    if (this._peekedCharacter !== undefined) {
      const character: ICharacter = this._peekedCharacter;
      this._peekedCharacter = undefined;
      return character;
    }

    if (this._endToken) {
      // already reached the end of the input
      return {
        value: EOI_CHARACTER,
        index: this._endToken.range.pos,
        line: this._endToken.line
      };
    }
    if (!this._cachedCurrentLine) {
      // Sanity check
      throw new Error('Tokenizer._currentLine should not be undeifned');
    }

    while (true) {
      if (this._bufferIndex < this._cachedCurrentLine.end) {
        return {
          value: this._cachedBuffer[this._bufferIndex],
          index: this._bufferIndex++,
          line: this._cachedCurrentLine
        };
      }

      // When we reach the logical end of line, we inject a "\n" character (since the
      // real EOL may be disembodied by the doc comment delimiters).
      // Since we don't want to move _bufferIndex, we need a little bit of extra state.
      if (!this._injectedNewline) {
        this._injectedNewline = true;
        return {
          value: '\n',
          index: this._bufferIndex,
          line: this._cachedCurrentLine
        };
      }
      this._injectedNewline = false;

      // Advance to the next line
      ++this._linesIndex;
      if (this._linesIndex >= this.lines.length) {
        // We advanced past the final line
        this._endToken = new Token(TokenKind.EndOfInput,
          this._cachedCurrentLine.getNewRange(this._cachedCurrentLine.end, this._cachedCurrentLine.end),
          this._cachedCurrentLine
        );
        return {
          value: EOI_CHARACTER,
          index: this._endToken.range.pos,
          line: this._endToken.line
        };
      }

      this._cachedCurrentLine = this.lines[this._linesIndex];
      this._bufferIndex = this._cachedCurrentLine.pos;
    }
  }

  /**
   * Similar to _getCharacter(), except it does not advance the (conceptual) input stream pointer.
   */
  private _peekCharacter(): ICharacter {
    if (this._peekedCharacter === undefined) {
      this._peekedCharacter = this._getCharacter();
    }
    return this._peekedCharacter;
  }
}
