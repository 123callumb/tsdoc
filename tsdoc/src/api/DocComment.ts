import { TextRange } from './TextRange';
import { ParseError } from './ParseError';

/**
 * Indicates the type of {@link DocNode}.
 */
export enum DocNodeKind {
  DocComment
}

/**
 * Represents a parsed documentation item.
 */
export abstract class DocNode {
  /**
   * Indicates the kind of DocNode.
   */
  public abstract readonly kind: DocNodeKind;

  public readonly range: TextRange;

  public constructor(range: TextRange) {
    this.range = range;
  }
}

/**
 * Constructor parameters for {@link DocComment}.
 */
export interface IDocCommentParameters {
  /** {@inheritdoc DocComment.sourceRange} */
  sourceRange: TextRange;

  /** {@inheritdoc DocComment.commentRange} */
  commentRange: TextRange;

  /** {@inheritdoc DocComment.lines} */
  lines: TextRange[];

  /** {@inheritdoc DocComment.parseErrors} */
  parseErrors: ParseError[];
}

/**
 * Represents a parsed documentation comment.
 */
export class DocComment extends DocNode {

  /** {@inheritdoc} */
  public readonly kind: DocNodeKind = DocNodeKind.DocComment;

  /**
   * Whereas {@link DocComment.range} tracks the start and end of the `/**` and `*\/` delimiters,
   * the `sourceRange` indicates the start and end of the original input that was parsed.
   */
  public readonly sourceRange: TextRange;

  /**
   * The text ranges corresponding to the lines of content inside the comment.
   */
  public readonly lines: TextRange[];

  /**
   * If any errors occurred during parsing, they are returned in this list.
   */
  public readonly parseErrors: ParseError[] = [];

  /**
   * Don't call this directly.  Instead use {@link TSDocParser}
   * @internal
   */
  public constructor(parameters: IDocCommentParameters) {
    super(parameters.commentRange);

    this.sourceRange = parameters.sourceRange;

    this.lines = parameters.lines;
    this.parseErrors = parameters.parseErrors;
  }
}
