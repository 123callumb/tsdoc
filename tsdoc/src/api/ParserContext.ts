import { TextRange } from './TextRange';
import { Token } from './Tokenizer';
import { ParseError } from './ParseError';

export class ParserContext {
  /**
   * The input text
   */
  public sourceRange: TextRange;

  /**
   * The range from the opening comment delimiter ("/**" to the closing comment delimiter.
   */
  public commentRange: TextRange;

  /**
   * The line ranges inside the doc comment.
   */
  public lines: TextRange[];

  /**
   * The list of tokens extracted from the lines.
   */
  public tokens: Token[];

  public parseErrors: ParseError[];

  public constructor(range: TextRange) {
    this.sourceRange = range;
    this.commentRange = TextRange.empty;
    this.lines = [];
    this.tokens = [];
    this.parseErrors = [];
  }

  public addError(range: TextRange, message: string, pos: number, end?: number): void {
    if (!end) {
      if (pos + 1 <= range.buffer.length) {
        end = pos + 1;
      } else {
        end = pos;
      }
    }
    this.parseErrors.push(
      new ParseError(message, range.getNewRange(pos, end))
    );
  }
}
