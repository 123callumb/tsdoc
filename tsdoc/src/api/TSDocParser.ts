import { Character } from '../internal/Character';

import { DocComment, IDocCommentParameters } from './nodes/DocComment';
import { TextRange } from './TextRange';
import { ParseError } from './ParseError';

// Internal parser state
enum State {
  // Initial state, looking for "/*"
  BeginComment1,
  // Looking for "*" or "* " after "/*"
  BeginComment2,
  // Like State.CollectingLine except immediately after the "/**"
  CollectingFirstLine,
  // Collecting characters until we reach a newline
  CollectingLine,
  // After a newline, looking for the "*" that begins a new line, or the "*/" to end the comment
  AdvancingLine,
  // Exiting the parser loop
  Done
}

/**
 * The main API for parsing TSDoc comments.
 */
export class TSDocParser {
  private static _addError(parameters: IDocCommentParameters, range: TextRange,
    message: string, pos: number, end?: number): void {
    if (!end) {
      if (pos + 1 <= range.buffer.length) {
        end = pos + 1;
      } else {
        end = pos;
      }
    }
    parameters.parseErrors.push(
      new ParseError(message, range.getNewRange(pos, end))
    );
  }

  public parseString(text: string): DocComment {
    return this.parseRange(TextRange.fromString(text));
  }

  public parseRange(range: TextRange): DocComment {
    const parameters: IDocCommentParameters = {
      range: TextRange.empty,
      sourceRange: range,
      lines: [],
      parseErrors: []
    };
    this._parseLines(parameters);

    return new DocComment(parameters);
  }

  /**
   * This step parses an entire code comment from slash-star-star until star-slash,
   * and extracts the content lines.  The lines are stored in IDocCommentParameters.lines
   * and the overall text range is assigned to IDocCommentParameters.range.
   */
  private _parseLines(parameters: IDocCommentParameters): void {
    const range: TextRange = parameters.sourceRange;
    const buffer: string = range.buffer;

    let commentRangeStart: number = 0;
    let commentRangeEnd: number = 0;

    // These must be set before entering CollectingFirstLine, CollectingLine, or AdvancingLine
    let collectingLineStart: number = 0;
    let collectingLineEnd: number = 0;

    let nextIndex: number = range.pos;
    let state: State = State.BeginComment1;

    while (state !== State.Done) {
      if (nextIndex >= range.end) {
        // reached the end of the input
        switch (state) {
          case State.BeginComment1:
          case State.BeginComment2:
            TSDocParser._addError(parameters, range, 'Expecting a "/**" comment', range.pos);
              return;
          default:
            TSDocParser._addError(parameters, range, 'Unexpected end of input', range.pos);
            return;
        }
      }

      const current: string = buffer[nextIndex];
      const currentIndex: number = nextIndex;
      ++nextIndex;
      const next: string = nextIndex < range.end ? buffer[nextIndex] : '';

      switch (state) {
        case State.BeginComment1:
          if (current === '/' && next === '*') {
            commentRangeStart = currentIndex;
            ++nextIndex; // skip the star
            state = State.BeginComment2;
          } else if (!Character.isWhitespace(current)) {
            TSDocParser._addError(parameters, range, 'Expecting a leading "/**"', nextIndex);
            return;
          }
          break;
        case State.BeginComment2:
          if (current === '*') {
            if (next === ' ') {
              ++nextIndex; // Discard the space after the star
            }
            collectingLineStart = nextIndex;
            collectingLineEnd = nextIndex;
            state = State.CollectingFirstLine;
          } else {
            TSDocParser._addError(parameters, range, 'Expecting a leading "/**"', nextIndex);
            return;
          }
          break;
        case State.CollectingFirstLine:
        case State.CollectingLine:
          if (current === '\n') {
            // Ignore an empty line if it is immediately after the "/**"
            if (state !== State.CollectingFirstLine || collectingLineEnd > collectingLineStart) {
              // Record the line that we collected
              parameters.lines.push(range.getNewRange(collectingLineStart, collectingLineEnd));
            }
            collectingLineStart = nextIndex;
            collectingLineEnd = nextIndex;
            state = State.AdvancingLine;
          } else if (current === '*' && next === '/') {
            if (collectingLineEnd > collectingLineStart) {
              parameters.lines.push(range.getNewRange(collectingLineStart, collectingLineEnd));
            }
            collectingLineStart = 0;
            collectingLineEnd = 0;
            ++nextIndex; // skip the slash
            commentRangeEnd = nextIndex;
            state = State.Done;
          } else if (!Character.isWhitespace(current)) {
            collectingLineEnd = nextIndex;
          }
          break;
        case State.AdvancingLine:
          if (current === '*') {
            if (next === '/') {
              collectingLineStart = 0;
              collectingLineEnd = 0;

              ++nextIndex; // skip the slash
              commentRangeEnd = nextIndex;
              state = State.Done;
            } else {
              // Discard the "*" at the start of a line

              if (next === ' ') {
                ++nextIndex; // Discard the space after the star
              }

              collectingLineStart = nextIndex;
              collectingLineEnd = nextIndex;
              state = State.CollectingLine;
            }
          } else if (current === '\n') {
            // Blank line
            parameters.lines.push(range.getNewRange(currentIndex, currentIndex));
            collectingLineStart = nextIndex;
          } else if (!Character.isWhitespace(current)) {
            // If the star is missing, then start the line here
            // Example: "/**\nL1*/"

            // (collectingLineStart was the start of this line)
            collectingLineEnd = currentIndex;
            state = State.CollectingLine;
          }
          break;
      }
    }

    parameters.range = range.getNewRange(commentRangeStart, commentRangeEnd);
  }
}
