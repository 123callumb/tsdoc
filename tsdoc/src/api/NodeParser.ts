import { ParserContext } from './ParserContext';
import { Token, TokenKind, Tokenizer } from './Tokenizer';
import { DocNode } from './nodes/DocNode';
import { DocPlainText } from './nodes/DocPlainText';
import { DocHtmlTag } from './nodes/DocHtmlTag';
import { DocBackslashEscape } from './nodes/DocBackslashEscape';
import { DocError } from './nodes/DocError';

export class NodeParser {
  private _tokens: Token[] = [];
  // The index into the _tokens array, of the current token being processed.
  private _tokenIndex: number = 0;

  /**
   * The output list of nodes that was parsed from the input tokens.
   */
  private _nodes: DocNode[] = [];

  /**
   * When we are accumulating a DocPlainText node, these are the tokens.
   */
  private _accumulatedPlainText: Token[];

  public constructor(parserContext: ParserContext) {
    this._tokens = parserContext.tokens;
    this._accumulatedPlainText = [];
  }

  public parse(): DocNode[] {

    while (true) {
      // Extract the next token
      const token: Token = this._tokens[this._tokenIndex];

      if (token.kind === TokenKind.EndOfInput) {
        break;
      }

      const nextToken: Token = this._tokens[this._tokenIndex + 1];

      switch (token.kind) {
        case TokenKind.Backslash:
          this._pushAccumulatedPlainText();

          // In CommonMark, a backslash is only allowed before a punctuation
          // character.  In all other context, the backslash is interpreted as a
          // literal character.
          if (Tokenizer.isPunctuation(nextToken.kind)) {
            this._nodes.push(new DocBackslashEscape({
              tokens: [ token, nextToken ]
            }));
            this._tokenIndex += 2; // extract both tokens
          } else {
            this._nodes.push(new DocError({
              tokens: [ token ]
            }));
            ++this._tokenIndex; // extract only the error token
          }
          break;
        case TokenKind.LessThan:
          this._pushAccumulatedPlainText();
          this._parseHtmlElement();
          break;
        default:
          // If nobody recognized this token, then accumulate plain text
          this._accumulatedPlainText.push(token);
          ++this._tokenIndex; // extract token
          break;
      }
    }
    this._pushAccumulatedPlainText();

    return this._nodes;
  }

  private _parseHtmlElement(): void {
    for (let i: number = this._tokenIndex; i < this._tokens.length; ++i) {
      const scannedToken: Token = this._tokens[i];
      if (scannedToken.kind === TokenKind.GreaterThan) {
        // Found the matching ">"
        this._nodes.push(
          new DocHtmlTag({
            tokens: this._tokens.slice(this._tokenIndex, i)
          })
        );
        this._tokenIndex = i + 1;
        return;
      }
    }
    // Otherwise it's an error
    this._nodes.push(new DocError({
      tokens: [ this._tokens[this._tokenIndex] ]
    }));
    ++this._tokenIndex;
  }

  private _pushAccumulatedPlainText(): void {
    if (this._accumulatedPlainText.length > 0) {
      this._nodes.push(
        new DocPlainText(
          {
            tokens: this._accumulatedPlainText
          }
        )
      );
      this._accumulatedPlainText = [];
    }
  }
}
