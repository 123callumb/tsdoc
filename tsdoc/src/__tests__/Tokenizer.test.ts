import { TSDocParser, DocComment } from '../index';
import { Tokenizer, Token, TokenKind } from '../api/Tokenizer';

function escape(s: string): string {
  return s.replace(/\n/g, '[n]')
    .replace(/\\/g, '[b]');
}

interface ISnapshotItem {
  indexOfLine: number;
  line: string;
  span: string;
  tokenKind: string;
}

function matchSnapshot(buffer: string): void {
  const tsdocParser: TSDocParser = new TSDocParser();
  const docComment: DocComment = tsdocParser.parseString(buffer);
  const tokenizer: Tokenizer = new Tokenizer(docComment.lines);

  const items: ISnapshotItem[] = [];
  const paddedSpace: string[]  = [ '',   ' ',  '  ',  '   ',  '    ' ];
  const paddedLArrow: string[] = [ '',   '>',  ' >',  '  >',  '   >' ];
  const paddedRArrow: string[] = [ '',   '<',  '< ',  '<  ',  '<   ' ];

  while (true) {
    const token: Token = tokenizer.getToken();

    let span: string = '';
    if (token.line.end > 0) {
      let i: number = token.line.pos - 1;
      while (i < token.range.pos - 1) {
        span += paddedSpace[escape(buffer[i]).length];
        ++i;
      }
      span += paddedLArrow[escape(buffer[i]).length];
      ++i;
      while (i < token.range.end) {
        span += paddedSpace[escape(buffer[i]).length];
        ++i;
      }
      if (i === token.line.end) {
        span += '<';
      } else {
        span += paddedRArrow[escape(buffer[i]).length];
        ++i;
        while (i < token.line.end) {
          span += paddedSpace[escape(buffer[i]).length];
          ++i;
        }

      }
    }

    items.push({
      indexOfLine: docComment.lines.indexOf(token.line),
      line: '>' + escape(token.line.toString()) + '<',
      span: span,
      tokenKind: TokenKind[token.kind]
    });

    if (token.kind === TokenKind.EndOfInput) {
      break;
    }
  }

  // Assert that getToken() should return EndOfInput repeatedly
  expect(tokenizer.getToken().kind).toEqual(TokenKind.EndOfInput);
  expect(tokenizer.getToken().kind).toEqual(TokenKind.EndOfInput);

  expect({
    buffer: escape(buffer),
    tokens: items
  }).toMatchSnapshot();
}

test('00 Tokenizer simple case', () => {
  matchSnapshot([
    '/**',
    ' * line 1 ', // extra space at end of line
    ' * line 2',
    ' */'
  ].join('\n'));
});

test('01 Tokenizer degenerate cases', () => {
  matchSnapshot('/***/');

  matchSnapshot([
    '/**',
    ' *',
    ' */'
  ].join('\n'));

  matchSnapshot([
    '/**',
    ' ',
    ' ',
    ' */'
  ].join('\n'));

});

test('02 Backslash escapes: positive examples', () => {
  matchSnapshot([
    '/**',
    ' * \\$\\@param',
    ' */'
  ].join('\n'));
});

test('03 Backslash escapes: negative examples', () => {
  matchSnapshot([
    '/**',
    ' * letter: \\A space: \\  end of line: \\',
    ' */'
  ].join('\n'));
});
