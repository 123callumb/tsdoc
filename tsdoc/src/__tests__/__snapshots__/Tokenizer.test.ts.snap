// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`00 Tokenizer simple case 1`] = `
Object {
  "buffer": "/**[n] * line 1 [n] * line 2[n] */",
  "tokens": Array [
    Object {
      "indexOfLine": 0,
      "line": ">line 1<",
      "span": ">      <",
      "tokenKind": "PlainText",
    },
    Object {
      "indexOfLine": 0,
      "line": ">line 1<",
      "span": "      ><",
      "tokenKind": "Newline",
    },
    Object {
      "indexOfLine": 1,
      "line": ">line 2<",
      "span": ">      <",
      "tokenKind": "PlainText",
    },
    Object {
      "indexOfLine": 1,
      "line": ">line 2<",
      "span": "      ><",
      "tokenKind": "Newline",
    },
    Object {
      "indexOfLine": 1,
      "line": ">line 2<",
      "span": "      ><",
      "tokenKind": "EndOfInput",
    },
  ],
}
`;

exports[`01 Tokenizer degenerate cases 1`] = `
Object {
  "buffer": "/***/",
  "tokens": Array [
    Object {
      "indexOfLine": -1,
      "line": "><",
      "span": "",
      "tokenKind": "EndOfInput",
    },
  ],
}
`;

exports[`01 Tokenizer degenerate cases 2`] = `
Object {
  "buffer": "/**[n] *[n] */",
  "tokens": Array [
    Object {
      "indexOfLine": 0,
      "line": "><",
      "span": "><",
      "tokenKind": "Newline",
    },
    Object {
      "indexOfLine": 0,
      "line": "><",
      "span": "><",
      "tokenKind": "EndOfInput",
    },
  ],
}
`;

exports[`01 Tokenizer degenerate cases 3`] = `
Object {
  "buffer": "/**[n] [n] [n] */",
  "tokens": Array [
    Object {
      "indexOfLine": 0,
      "line": "><",
      "span": "><",
      "tokenKind": "Newline",
    },
    Object {
      "indexOfLine": 1,
      "line": "><",
      "span": "><",
      "tokenKind": "Newline",
    },
    Object {
      "indexOfLine": 1,
      "line": "><",
      "span": "><",
      "tokenKind": "EndOfInput",
    },
  ],
}
`;

exports[`02 Backslash escapes: positive examples 1`] = `
Object {
  "buffer": "/**[n] * [b]$[b]@param[n] */",
  "tokens": Array [
    Object {
      "indexOfLine": 0,
      "line": ">[b]$[b]@param<",
      "span": ">    <        ",
      "tokenKind": "BackslashEscapedCharacter",
    },
    Object {
      "indexOfLine": 0,
      "line": ">[b]$[b]@param<",
      "span": "    >    <    ",
      "tokenKind": "BackslashEscapedCharacter",
    },
    Object {
      "indexOfLine": 0,
      "line": ">[b]$[b]@param<",
      "span": "        >     <",
      "tokenKind": "PlainText",
    },
    Object {
      "indexOfLine": 0,
      "line": ">[b]$[b]@param<",
      "span": "             ><",
      "tokenKind": "Newline",
    },
    Object {
      "indexOfLine": 0,
      "line": ">[b]$[b]@param<",
      "span": "             ><",
      "tokenKind": "EndOfInput",
    },
  ],
}
`;

exports[`03 Backslash escapes: negative examples 1`] = `
Object {
  "buffer": "/**[n] * letter: [b]A space: [b]  end of line: [b][n] */",
  "tokens": Array [
    Object {
      "indexOfLine": 0,
      "line": ">letter: [b]A space: [b]  end of line: [b]<",
      "span": ">                                         <",
      "tokenKind": "PlainText",
    },
    Object {
      "indexOfLine": 0,
      "line": ">letter: [b]A space: [b]  end of line: [b]<",
      "span": "                                         ><",
      "tokenKind": "Newline",
    },
    Object {
      "indexOfLine": 0,
      "line": ">letter: [b]A space: [b]  end of line: [b]<",
      "span": "                                         ><",
      "tokenKind": "EndOfInput",
    },
  ],
}
`;
